### ISSUE TITLE
Unit test fails in unexpected way when dynamically creating component

### ISSUE THREAD

**Issue by RubMertens** [angular/angular#41083]
# 🐞 bug report

### Affected Package

Probably caused by an issue in `TestBed`.
@angular/core/testing

### Is this a regression?
I have not tested in a different version, so I do not know.

### Description

 A clear and concise description of the problem...

We have a component that functions as a facade for some other components. When I introduced a custom pipe to one of the components generated at runtime, the tests started failing in a strange way. 

## 🔬 Minimal Reproduction

 

https://github.com/RubMertens/testing-pipe-repro 

## 🔥 Exception or Error
<pre><code>
 Failed: NG0302: The pipe 'displayValue' could not be found!. Find more at https://angular.io/errors/NG0302
        error properties: Object({ code: '302' })
        Error: NG0302: The pipe 'displayValue' could not be found!. Find more at https://angular.io/errors/NG0302
            at getPipeDef$1 (node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:25738:1)
            at ɵɵpipe (node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:25695:1)        
            at UsesPipeComponent_Template (ng:///UsesPipeComponent.js:7:9)
            at executeTemplate (node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:9544:1)
            at renderView (node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:9351:1)     
            at renderComponent (node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:10626:1)
            at renderChildComponents (node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:9216:1)
            at renderView (node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:9376:1)     
            at ComponentFactory$1.create (node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:25079:1)
            at ViewContainerRef.createComponent (node_modules/@angular/core/__ivy_ngcc__/fesm2015/core.js:23118:1)
Chrome 88.0.4324.190 (Windows 10): Executed 1 of 1 (1 FAILED) (0.082 secs / 0.074 secs)
</code></pre>

## 🌍  Your Environment

**Angular Version:**
<pre><code>

     _                      _                 ____ _     ___ 
    / \   _ __   __ _ _   _| | __ _ _ __     / ___| |   |_ _|
   / △ \ | '_ \ / _` | | | | |/ _` | '__|   | |   | |    | | 
  / ___ \| | | | (_| | |_| | | (_| | |      | |___| |___ | | 
 /_/   \_\_| |_|\__, |\__,_|_|\__,_|_|       \____|_____|___|
                |___/
    

Angular CLI: 11.2.3
Node: 14.16.0      
OS: win32 x64      

Angular: 11.2.4
... animations, common, compiler, compiler-cli, core, forms
... platform-browser, platform-browser-dynamic, router
Ivy Workspace: Yes

Package                         Version
---------------------------------------------------------
@angular-devkit/architect       0.1102.3
@angular-devkit/build-angular   0.1102.3
@angular-devkit/core            11.2.3
@angular-devkit/schematics      11.2.3
@angular/cli                    11.2.3
@schematics/angular             11.2.3
@schematics/update              0.1102.3
rxjs                            6.6.6
typescript                      4.1.5
</code></pre>

**Anything else relevant?**

It can be fixed by declaring the Child view component in the TestBed of the tests.
I would like to understand why the error talks about the pipe used by a child component that should not be known in the 

TestBed?

**Comment by petebacondarwin** [comment]
I suspect that the problem is that because you have not included the `UsesPipeComponent` in the testing module, that component does not know about the `displayValue` pipe.  If you think about it you have a testing module that contains:

```ts
TestBed.configureTestingModule({
  declarations: [
    AppComponent,
    DisplayValuePipe,
  ]
})
```

So the `AppComponent` and the `DisplayValuePipe` are in the same "module scope" so they know about each other.

If you don't declare the `UsesPipeComponent` in the testing module then it is effectively a "standalone component" that has no knowledge of other components and pipes.

When you add `UsesPipeComponent` to the testing module, it joins their module scope and so is aware of the `displayValue` pipe.

---

I would say this is working as expected, and I am not sure how we could make the error more helpful, since there is no way that the runtime could know how you have configured the rest of the system.

**Comment by RubMertens** [comment]
@petebacondarwin Thank you for the swift reply!

The way you explained it made sense to me, thank you! It is as if I did not declare the `UsesPipeComponent`  at all in the AppModule. In fact, when I remove it from my actual `AppModule` I see the same error in the browser as I did in the test.

It was a difficult error to track down, but I understand it is hard to give a more clear error message.

I will be closing the issue and again, thank you for your feedback! 😄
